### PR TITLE
eni: Add garbage collector for leaked ENIs

### PR DESCRIPTION
**Pull request by gandro**
This PR adds a new garbage collector for leaked ENIs. ENIs might be
leaked if the operator is unable to perform an ENI delete operation, for
example because the EC2 API is (temporarily) unavailable or the operator
crashed before it was able to issue the delete operation.

The GC runs in a regular interval and removes any ENIs with a predefined
set of tags. The tags are added to newly created ENIs, meaning this GC
does not remove any ENIs created by older versions of Cilium. This is
done to avoid removing any ENIs which were never managed by Cilium.

The tags used for ENI creation and deletion, as well as the GC interval,
can both be configured using operator flags. If the GC interval is set
to zero, the ENI GC is disabled.

**Review per commit**

### CODE REVIEW

**Comment by gandro** [comment]
Besides the unit tests, I have also tested this in a real cluster with a small test of ENIs:

```
# Run 1
2022-09-22T16:32:58.789403029Z level=debug msg="Marked unattached interfaces for garbage collection" numInterfaces=3 subsys=eni

# Run 2 - Note: I forced a race manually, by attaching one of the ENIs marked 
# for garbage collection to an instance. As you can see, the API did not remove the in-use ENI:
2022-09-22T16:37:58.790004387Z level=debug msg="Garbage collecting ENI" eniID=eni-03fc53ab5ad26d24e subsys=eni
2022-09-22T16:37:59.016210891Z level=debug msg="Failed to garbage collect ENI" error="operation error EC2: DeleteNetworkInterface, https response error StatusCode: 400, RequestID: fa502304-0dbc-403f-bbe2-c30d2615bfc1, api error InvalidParameterValue: Network interface 'eni-03fc53ab5ad26d24e' is currently in use." subsys=eni
2022-09-22T16:37:59.016280762Z level=debug msg="Garbage collecting ENI" eniID=eni-02714735b24b4dfd1 subsys=eni
2022-09-22T16:37:59.343633394Z level=debug msg="Garbage collecting ENI" eniID=eni-01d64d659b7116851 subsys=eni

# Run 3 & 4
# note: I unattached a ENI manually from an existing instance
2022-09-22T16:42:59.921815277Z level=debug msg="Marked unattached interfaces for garbage collection" numInterfaces=1 subsys=eni
2022-09-22T16:47:59.922109406Z level=debug msg="Garbage collecting ENI" eniID=eni-027e57fe40b35535e subsys=eni
```

**Comment by gandro** [comment]
Last push fixed a bug in the ec2shim around pagination. MaxResults did not work as I expected it to.

```diff
diff --git a/pkg/aws/ec2/ec2.go b/pkg/aws/ec2/ec2.go
index e4f7a95931ea..d90a8b87bec0 100644
--- a/pkg/aws/ec2/ec2.go
+++ b/pkg/aws/ec2/ec2.go
@@ -170,6 +170,9 @@ func (c *Client) GetDetachedNetworkInterfaces(ctx context.Context, tags ipamType
                for _, eni := range output.NetworkInterfaces {
                        result = append(result, aws.ToString(eni.NetworkInterfaceId))
                }
+               if len(result) >= int(maxResults) {
+                       break
+               }
        }
        return result, nil
 }
```

**Review by joestringer — APPROVED** [review]
The docs changes look trivially correct. Please re-request my review if you'd like me to take a closer look at the logic, otherwise I will leave it to the corresponding codeowners.

**Review by pippolo84 — APPROVED** [review]
LGTM :+1:

**Comment by gandro** [comment]
- [net-next](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.25-kernel-net-next/291/testReport/junit/Suite-k8s-1/25/K8sDatapathConfig_Check_BPF_masquerading_with_ip_masq_agent_VXLAN/) hit #21120 - unrelated known flake

File: install/kubernetes/cilium/values.yaml
Comment (squeed) [inline install/kubernetes/cilium/values.yaml]: Is there ever a chance that we could accidentally select ENIs owned by a different cluster? Should we tag this with a cluster name?

Comment (gandro) [reply install/kubernetes/cilium/values.yaml]: Yes, we will very likely collect ENIs owned by different clusters. I don't consider this a huge problem (the mechanism is explicitly designed to not need synchronization). But if users want/need to turn off GC for a particular cluster, they will have to either use per-cluster tags here, or set a global subnet tag filter unique to each cluster.

Unfortunately, I expect most users to not have a clustername set. `cilium install` does set one automatically, but most advanced users will use Helm and thus will have an empty clustername. But maybe we should still automatically add a cluster tag here _if_ the clustername is set and recommend it in the docs for ENI IPAM based installations.

**Review by sayboras — APPROVED** [review]
I don't have other comment :100:

File: install/kubernetes/cilium/values.yaml
Comment (squeed) [reply install/kubernetes/cilium/values.yaml]: Given that this only collects detached ENIs (correct me if I'm wrong), I assume it's not a big deal if we GC ENIs not "owned" by us, right? If so, this discussion is mostly academic.

But could we theoretically step on an ENI in the creation process?

**Review by christarazi — APPROVED** [review]

File: install/kubernetes/cilium/values.yaml
Comment (christarazi) [reply install/kubernetes/cilium/values.yaml]: @squeed I believe this is handled by the GC logic waiting for the next cycle to GC ENIs. Basically, it's like a mark and sweep GC. First round we collect all the ENIs (mark) and the next round we delete (sweep). Please correct me if I'm wrong @gandro.

Comment (gandro) [reply install/kubernetes/cilium/values.yaml]: @christarazi Yes, that's correct. Given an offline discussion with @squeed I will try to add a best-effort clustername tag as well, we might even be able to extract the EKS clustername from the EC2 API where available.

While generally speaking I don't think it is a problem if we GC ENIs from other clusters, if GC misbehaves and the user wants to turn it off or reconfigure, then cross-cluster GC will become an issue (since the settings in a different cluster will affect the GC in the local cluster and vice-versa).

Comment (gandro) [reply install/kubernetes/cilium/values.yaml]: @squeed @christarazi I have updated the PR to auto-detect the clustername and tag ENIs with the clustername if available. Please take a look.

**Review by squeed — APPROVED** [review]

**Review by christarazi — APPROVED** [review]

**Comment by gandro** [comment]
Marking ready to merge. The code added in this PR is only active in ENI mode, which passed twice. Full Jenkins CI also [passed previously](https://github.com/cilium/cilium/pull/21409#issuecomment-1258335832) and I've also tested the code (including the new clustername fallback) manually as well.

**Comment by gandro** [comment]
Missed a bunch of missing copyright headers. Nothing major, but I want to add them before we merge.

**Comment by gandro** [comment]
Rebased, fixed the missing copyright header and added a small doc section about the cluster name detection. Re-running full CI on the rebased branch.

**Pull request merged**
